soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Select.scala

```diff
@@ -22,33 +19,109 @@ case class SubSelect(selects: BinaryTree[Select], alias: String)
+    // ok we're going to rewrite
+    //   join @foo(x, y, z) as bleh
+    // to
+    //   join (select @expr.:*, @expr.* from (select x,y,z from @single_row) as vars join lateral (body) as expr on true) as bleh;
+    // ..or that's the ideal, anyway.  Turns out you can't FROM a
+    // query in soql, so we'll actually rewrite to "..from @single_row
+    // join (select x,y,z) join lateral (body)" instead.
+    //
+    // Probably will try to turn `from single_row join bleh on true`
+    // into `from bleh` over in the postgres adapter later.
+    if(udf.arguments.length != params.length) {
+      throw MismatchedParameterCount(expected = udf.arguments.length, got = params.length)(position)
     }
+    val expr = aliasProvider("_expr")
+    val vars = aliasProvider("_vars")
+    val singleRowFrom = Some(TableName(TableName.SingleRow))
+    val labelledJoin =
+      Select(
+        distinct = false,
+        selection = Selection(Some(StarSelection(Some(expr), Nil)),
+                              Seq(StarSelection(Some(expr), Nil)),
+                              Nil),
+        from = singleRowFrom,
+        joins = Seq(
+          InnerJoin(
+            JoinQuery(
+              Leaf(
+                Select(
+                  distinct = false,
+                  selection = Selection(None, Nil,
+                                        udf.arguments.zip(params).map {
+                                          case ((holeName, typ), expression) =>
+                                            SelectedExpression(FunctionCall(SpecialFunctions.Cast(typ), Seq(expression), None)(NoPosition, NoPosition),
+                                                               Some((ColumnName(holeName.name), NoPosition)))
+                                        }.toList),
+                  from = singleRowFrom,
+                  joins = Nil, where = None, groupBys = Nil, having = None, orderBys = Nil, limit = None, offset = None, search = None)),
+              vars),
+            on = BooleanLiteral(true)(NoPosition),
+            lateral = false),
+          InnerJoin(
+            JoinQuery(
+              Select.walkTreeReplacingHoles(udf.body, { hole =>
+                ColumnOrAliasRef(Some(vars), ColumnName(hole.name.name))(hole.position)
+              }),
+              expr),
+            on = BooleanLiteral(true)(NoPosition),
+            lateral = true)),
+        where = None, groupBys = Nil, having = None, orderBys = Nil, limit = None, offset = None, search = None)
 
-    List(Some(subSelectStr), itrToString("AS", aliasStrOpt.map(TableName.removeValidPrefix))).flatString
+    JoinQuery(Leaf(labelledJoin), tableName.alias.getOrElse(tableName.name))
+  }
+}
+
+trait AliasProvider {
+  def apply(base: String): String
+}
+object AliasProvider {
+  implicit val instance = new AliasProvider {
+    def apply(base: String) = base + UUID.randomUUID().toString.replaceAll("-", "_")
```

Comment:
randomUUID seems to generate different soql seen by soda fountain.  It seems that a stable generator might be better and has less chance to have fallouts?